server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/PushDataHandler.scala

```diff
@@ -0,0 +1,406 @@
+    val exception = fileWriter.getException
+    if (exception != null) {
+      logWarning(s"[handlePushData] fileWriter $fileWriter has Exception $exception")
+      val message = if (isMaster) {
+        StatusCode.PushDataFailMain.getMessage()
+      } else {
+        StatusCode.PushDataFailSlave.getMessage()
+      }
+      callback.onFailure(new Exception(message, exception))
+      return
+    }
+    if (isMaster && fileWriter.getFileLength > fileWriter.getSplitThreshold()) {
+      fileWriter.setSplitFlag()
+      if (fileWriter.getSplitMode == PartitionSplitMode.soft) {
+        callback.onSuccess(ByteBuffer.wrap(Array[Byte](StatusCode.SoftSplit.getValue)))
+      } else {
+        callback.onSuccess(ByteBuffer.wrap(Array[Byte](StatusCode.HardSplit.getValue)))
+        return
+      }
+    }
+    fileWriter.incrementPendingWrites()
+
+    // for master, send data to slave
+    if (location.getPeer != null && isMaster) {
+      pushData.body().retain()
+      replicateThreadPool.submit(new Runnable {
+        override def run(): Unit = {
+          val peer = location.getPeer
+          val peerWorker = new WorkerInfo(peer.getHost, peer.getRpcPort, peer.getPushPort,
+            peer.getFetchPort, peer.getReplicatePort, -1, null)
+          if (unavailablePeers.containsKey(peerWorker)) {
+            pushData.body().release()
+            wrappedCallback.onFailure(new Exception(s"Peer $peerWorker unavailable!"))
+            return
+          }
+          try {
+            val client = pushClientFactory.createClient(peer.getHost, peer.getReplicatePort,
+              location.getReduceId)
+            val newPushData = new PushData(
+              PartitionLocation.Mode.Slave.mode(),
+              shuffleKey,
+              pushData.partitionUniqueId,
+              pushData.body)
+            client.pushData(newPushData, wrappedCallback)
+          } catch {
+            case e: Exception =>
+              pushData.body().release()
+              unavailablePeers.put(peerWorker, System.currentTimeMillis())
+              wrappedCallback.onFailure(e)
+          }
+        }
+      })
+    } else {
+      wrappedCallback.onSuccess(ByteBuffer.wrap(Array[Byte]()))
+    }
+
+    try {
+      fileWriter.write(body)
+    } catch {
+      case e: AlreadyClosedException =>
+        fileWriter.decrementPendingWrites()
+        val (mapId, attemptId) = getMapAttempt(body)
+        val endedAttempt = if (shuffleMapperAttempts.containsKey(shuffleKey)) {
+          shuffleMapperAttempts.get(shuffleKey)(mapId)
+        } else -1
+        logWarning(s"Append data failed for task(shuffle $shuffleKey, map $mapId, attempt" +
+          s" $attemptId), caused by ${e.getMessage}")
+      case e: Exception =>
+        logError("Exception encountered when write.", e)
+    }
+  }
+
+  def handlePushMergedData(
+        pushMergedData: PushMergedData,
+        callback: RpcResponseCallback): Unit = {
```

Comment:
4 identation

Reply:
As for indentations, I get an idea. I will modify check style and scala style configurations to align indentations. RSS uses 2 spaces as indentations, for now, it's the same as the scala style guide (https://docs.scala-lang.org/style/indentation.html#line-wrapping).

Reply:
> 

👍